cmd/acr/purge.go

```diff
@@ -568,38 +586,48 @@ func dryRunPurge(ctx context.Context, acrClient api.AcrCLIClientInterface, login
-		// This will act as a set if a key is present then it should not be deleted because it is referenced by a multiarch manifest
+		// This will act as a set if a key is present then it should not be deleted
+		// because it is referenced by a multiarch manifest or a manifest has subject attached
 		// that will not be deleted
-		doNotDelete := map[string]bool{}
+		doNotDelete := make(map[string]struct{})
 		candidatesToDelete := []acr.ManifestAttributesBase{}
 		// Iterate over all manifests to discover multiarchitecture manifests
 		for resultManifests != nil && resultManifests.ManifestsAttributes != nil {
 			manifests := *resultManifests.ManifestsAttributes
 			for _, manifest := range manifests {
-				// If the manifest is manifest list and would not get deleted then mark it's dependant manifests as not deletable.
-				if (*manifest.MediaType == manifestListContentType || *manifest.MediaType == manifestIndexContentType) && (*countMap)[*manifest.Digest] != deletedTags[*manifest.Digest] {
-					var manifestListBytes []byte
-					manifestListBytes, err = acrClient.GetManifest(ctx, repoName, *manifest.Digest)
-					if err != nil {
-						return -1, -1, err
-					}
-					var manifestList multiArchManifest
-					err = json.Unmarshal(manifestListBytes, &manifestList)
-					if err != nil {
-						return -1, -1, err
-					}
-					for _, dependentDigest := range manifestList.Manifests {
-						doNotDelete[dependentDigest.Digest] = true
+				if (*countMap)[*manifest.Digest] != deletedTags[*manifest.Digest] {
+					if _, ok := mediaTypes[*manifest.MediaType]; ok {
+						manifestBytes, err := acrClient.GetManifest(ctx, repoName, *manifest.Digest)
+						if err != nil {
+							return -1, -1, err
+						}
+						var customManifest customManifest
+						err = json.Unmarshal(manifestBytes, &customManifest)
+						if err != nil {
+							return -1, -1, err
+						}
+						if customManifest.Subject != nil {
+							// if manifest has subject, manifest is marked to not be deleted
+							doNotDelete[*manifest.Digest] = struct{}{}
+						} else if manifest.Tags == nil {
+							// If the manifest has no subject or tag, it is a candidate for deletion
+							candidatesToDelete = append(candidatesToDelete, manifest)
+						}
+						if len(customManifest.Manifests) > 0 {
+							for _, dependentManifest := range customManifest.Manifests {
+								doNotDelete[dependentManifest.Digest] = struct{}{}
+							}
+						}
 					}
-				} else if (*countMap)[*manifest.Digest] == deletedTags[*manifest.Digest] {
+				} else {
 					// If the manifest has the same amount of tags as the amount of tags deleted then it is a candidate for deletion.
 					candidatesToDelete = append(candidatesToDelete, manifest)
 				}
-			}
-			lastManifestDigest = *manifests[len(manifests)-1].Digest
-			resultManifests, err = acrClient.GetAcrManifests(ctx, repoName, "", lastManifestDigest)
-			if err != nil {
-				return -1, -1, err
+				lastManifestDigest = *manifests[len(manifests)-1].Digest
```

Comment:
Nit. This part is to get more manifests for the second part. It seems like an isolated piece of functionality; I would suggest spacing it into its own block and noting the functionality.

Reply:
done. added space and comments.